prov/mxm/src/mlxm_cq.c

```diff
@@ -0,0 +1,228 @@
+        struct fi_context *ctx;
+        struct fi_cq_err_entry *cqe =
+                (struct fi_cq_err_entry *) buf;
+
+        fid_cq = container_of(cq, mlxm_fid_cq_t, cq);
+
+        if (!fid_cq->err_q.head) {
+                return 0;
+        }
+
+
+
+        MLXM_CQ_DEQUEUE(fid_cq->err_q, ctx);
+
+        mlxm_req = ctx->internal[1];
+        cqe->op_context	= ctx;
+        cqe->flags	= FI_TAGGED;
+        cqe->buf = NULL;
+        if ((uint64_t)(ctx->internal[3]) == FI_SEND) {
+                mxm_sreq = &mlxm_req->mxm_req.sreq;
+                assert(mxm_sreq->base.error != MXM_OK);
+                cqe->prov_errno = mxm_sreq->base.error;
+                cqe->len	= mxm_sreq->base.data.buffer.length;
+                cqe->olen       = 0;
+                cqe->data	= mxm_sreq->op.send.imm_data;
+                cqe->tag	= mxm_sreq->op.send.tag;
+                cqe->flags |= FI_SEND;
+
+        } else {
+                mxm_rreq = &mlxm_req->mxm_req.rreq;
+
+                assert(mxm_rreq->base.error != MXM_OK);
+
+                cqe->prov_errno = mxm_sreq->base.error;
+                cqe->len	= mxm_rreq->completion.actual_len;
+                cqe->olen       = mxm_rreq->completion.sender_len -
+                        cqe->len;
+                cqe->data	= mxm_rreq->completion.sender_imm;
+                cqe->tag	= mxm_rreq->completion.sender_tag;
+                cqe->flags |= FI_RECV;
+        }
+        cqe->err = -mlxm_errno(cqe->prov_errno);
+        cqe->tag |= (((uint64_t)mlxm_req->mq_id) << 32);
+
+
+
+
+        MPOOL_RETURN(mlxm_globals.req_pool, struct mlxm_req, mlxm_req);
+        return 1;
+}
+
+
+
+static const char *mlxm_cq_strerror(struct fid_cq *cq, int prov_errno,
+                                    const void *prov_data, char *buf, size_t len)
+{
+	return mxm_error_string(prov_errno);
+}
+
+
+static int mlxm_cq_control(fid_t fid, int command, void *arg)
+{
+	return -ENOSYS;
+}
+
+static struct fi_ops mlxm_fi_ops = {
+	.size    = sizeof(struct fi_ops),
+	.close   = mlxm_cq_close,
+        .control = mlxm_cq_control,
+};
+
+static struct fi_ops_cq mlxm_cq_ops = {
+	.size     = sizeof(struct fi_ops_cq),
+	.readfrom = mlxm_cq_readfrom,
+	.read     = mlxm_cq_read,
+	.readerr  = mlxm_cq_readerr,
+        .strerror = mlxm_cq_strerror,
+};
+
+int mlxm_cq_open(struct fid_domain *domain, struct fi_cq_attr *attr,
+                 struct fid_cq **cq, void *context)
+{
+        mlxm_fid_cq_t 	*fid_cq;
+
+        fid_cq = (mlxm_fid_cq_t *) calloc(1, sizeof *fid_cq);
+	if (!fid_cq)
+		return -ENOMEM;
+
+        fid_cq->cq.fid.fclass	= FI_CLASS_CQ;
+	fid_cq->cq.fid.context	= context;
+	fid_cq->cq.fid.ops	        = &mlxm_fi_ops;
+	fid_cq->cq.ops		= &mlxm_cq_ops;
+        fid_cq->mxm_context     = mlxm_globals.mxm_context;
+        fid_cq->ok_q.head = NULL;
+        fid_cq->ok_q.tail = NULL;
+        fid_cq->err_q.head = NULL;
+        fid_cq->err_q.tail = NULL;
+
+	*cq = &fid_cq->cq;
+	return 0;
+}
```

Comment:
There's a lot of extra white space inserted between lines and code alignment issues.  Please use tabs with 8-character alignment throughout the code.